### PR TITLE
Allow k8s specific pod config

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

We now allow charms to specify some k8s specific pod config when sending the pod spec yaml to juju.
The allowed extra k8s config includes:
- livenessProbe
- readinessProbe
- imagePullPolicy

The solution involved removing the hard coded k8s spec parsing and making it determined by the provider. In turn, this required enhancing the caas broker to have a Provider() method similar to what we use for cloud providers. The changes were threaded through the code as needed.

## QA steps

Modify the caas charms and deploy on top of k8s

NB - this recently introduced upstream k8s regression prevents full QA as it stops the operator from deploying https://github.com/kubernetes/kubernetes/issues/61076

### CODE REVIEW

**Comment by wallyworld** [comment]
[CAAS provider type - remove k8 hard code](https://trello.com/c/S8Tb9jyN/31-caas-provider-type-remove-k8-hard-code)

**Comment by wallyworld** [comment]
[pass-through of native pod config](https://trello.com/c/RYrX9jeo/55-pass-through-of-native-pod-config)

File: caas/containers.go [hunk 1]
```diff
@@ -3,67 +3,77 @@
...
+import "github.com/juju/errors"
 
 // FileSet defines a set of files to mount
 // into the container.
 type FileSet struct {
-	Name      string            `yaml:"name"`
-	MountPath string            `yaml:"mount-path"`
-	Files     map[string]string `yaml:"files"`
+	Name      string            `yaml:"name" json:"name"`
+	MountPath string            `yaml:"mountPath" json:"mountPath"`
```
Comment (anastasiamac) [inline caas/containers.go]: I wonder why camelCase change though? Why not hyphenated?

File: caas/containers.go [hunk 2]
```diff
@@ -3,67 +3,77 @@
...
-		return nil, errors.New("require at least one pod spec")
+	if spec.Image == "" {
+		return errors.New("spec image is missing")
 	}
-	for _, container := range spec.Containers {
-		if container.Name == "" {
-			return nil, errors.New("spec name is missing")
+	for _, fs := range spec.Files {
+		if fs.Name == "" {
+			return errors.New("file set name is missing")
```
Comment (anastasiamac) [inline caas/containers.go]: This is iterating over a few files and if one name is missing, we will bail? Did you intend that or did you actually want to log and 'continue'? 
The same goes for mount path check below..

File: caas/containers.go (hunk 1 above)
Comment (wallyworld) [reply caas/containers.go]: This is what kubernetes uses so I wanted to be consistent with what people expect to use already

File: caas/containers.go (hunk 2 above)
Comment (wallyworld) [reply caas/containers.go]: Intended to bail because even one missing name messes up the entire config

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**